### PR TITLE
Docker

### PR DESCRIPTION
**Pull request by nam20485**
Docker container for frontend react service using yarn. 

`bin/build.sh -p|-d` and `bin/start.sh -p|-d` work, much like the backend containers. They can be used to run the web service or ignored and existing developer process can be used. Container is primarily for the travis CI/Google cloud pipeline.

### CODE REVIEW

**Comment by marcaaron** [comment]
I'm unsure of how to proceed in testing this and also a little unsure about what the exact changes are. Would you be so kind as to provide more information about what's needed from us in reviewing this PR?

Main question I have is... does this change affect the local dev environment or strictly relate to deployment? I'm assuming we can just run our various yarn scripts and ignore the docker bits without docker running? Thanks :)

File: scripts/fetch_gae_sdk.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python
```
Comment (apodiptych) [inline scripts/fetch_gae_sdk.py]: How does Google App Engine fit into all this? Since it's a "serverless" platform, I imagine this isn't portable to a standard VPS somewhere else?

File: wobbly-frontend.code-workspace [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (apodiptych) [inline wobbly-frontend.code-workspace]: Is this file necessary?

File: docker-compose-production.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    build:
+      context: .
+      dockerfile: DOCKERFILE
+    image: $DOCKER_IMAGE
+    command: bash bin/docker-entrypoint.sh
+    volumes:
+      - .:/code
+      - /code/node_modules
+    ports:
+      - "8080:8080"
```
Comment (apodiptych) [inline docker-compose-production.yml]: `yarn web` just runs a local dev server that isn't suitable for production -- for that, you'd want to run `yarn build` and then serve the files in `public/`.

**Comment by nam20485** [comment]
>I'm unsure of how to proceed in testing this and also a little unsure about what the exact changes are. Would you be so kind as to provide more information about what's needed from us in reviewing this PR?

>Main question I have is... does this change affect the local dev environment or strictly relate to deployment? I'm assuming we can just run our various yarn scripts and ignore the docker bits without docker running? Thanks :)

@marcaaron You are correct. It does not affect the local dev environment, the purpose is to enable docker deployment in our CI pipeline. In fact it should not affect anything (if it does please let me know and I will fix it), unless you elect to run one of the scripts I provided. Your existing local dev process should be intact. It does provide a travis config file, docker config file, and various scripts and support files, which provide the ability to build, run, and unit test the docker image.

Please let me know if you have further questions or want further clarification.

File: scripts/fetch_gae_sdk.py (hunk 1 above)
Comment (nam20485) [reply scripts/fetch_gae_sdk.py]: Google App Engine doesn't fit in to the pipeline per se. That script is the mechanism by which I install the Google Cloud SDK command line client, which is used in my script `deploy-gcloud.sh`.

e.g.,
```
gcloud auth activate-service-account --key-file client-secret.json
gcloud config set project wobbly-backend
gcloud auth print-access-token | docker login -u oauth2accesstoken --password-stdin $DOCKER_REPO
```

I use the Google Cloud SDK client to authenticate to our Google Cloud project so that I can push the docker image into our container registry (basically a docker image repository). 

Google App Engine is one choice for deployment and running of our frontend service docker image, but I will probably end up using Google Compute Engine, which is tailored for running docker images. As you mention, this Google Cloud pipeline is a managed/"serverless" platform, not a self-managed VPS solution. I would like to enable this pipeline until we settle on what hosting solution we want to go forward with, at which point I will need to adapt some portions of this pipeline to deploy to that platform. However, this is a quick way to get our CI pipeline up and running and develop the overall framework.

File: wobbly-frontend.code-workspace (hunk 2 above)
Comment (nam20485) [reply wobbly-frontend.code-workspace]: That is up to you. I check in the VSCode workspace file (its pretty much empty) for my backend to enable a quick method of opening the source code in a modern IDE (MS VSCode), since this is a good IDE for new users. If you want, I can remove it from the repo and add it to .gitignore.

File: docker-compose-production.yml (hunk 3 above)
Comment (nam20485) [reply docker-compose-production.yml]: OK sorry, I don't have much frontend/react/yarn/node experience, and was just trying to get the docker image to run.

It sounds like `yarn web` is OK for running the service for local development, but I will need to replace it in the production docker image with whatever we want to use to serve the app with in production. Do we know what we want to use for serving the production service?

**Comment by marcaaron** [comment]
@nam20485 Ok thanks! So, one more thing... we're deploying the web version to GCP? And will docker handle deployment for mobile builds? I'm less knowledgeable in that area so curious if docker will help out with that once we have a mobile app to ship. But maybe that's a convo for another day.

File: scripts/fetch_gae_sdk.py (hunk 1 above)
Comment (marcaaron) [reply scripts/fetch_gae_sdk.py]: I think it would be helpful to have an in progress version of the web build up. If it's easy enough to do that with GCE then I think that can work for now.

**Comment by nam20485** [comment]
Yes, the current, initial version deploys web version to GCP. Once we decide where we really want to host the services, I will modify it to deploy to that. (Tao  has mentioned using a VPS provider). 

As far as the mobile app build and deploy, yes that will need to a separate future conversation. 

I fully intend to add additional functionality to build, package, and deploy the iOS version. But I have no experience doing so, so I will need to research and develop that part of  the pipeline. Or if someone else comes along who knows it they could possibly add it.

File: wobbly-frontend.code-workspace (hunk 2 above)
Comment (nam20485) [reply wobbly-frontend.code-workspace]: I noticed that you already included the `.vscode` directory which contains VS Code default settings. so it seems only natural to include a workspace file also.

File: scripts/fetch_gae_sdk.py (hunk 1 above)
Comment (nam20485) [reply scripts/fetch_gae_sdk.py]: @marcaaron The quickest and easiest path to having an in progress version of the web build up is indeed using GCE, for the simple fact that I already have 90% of a GCE pipeline working. All I need at this point is to have this PR merged in and then I can create the last stage.

File: wobbly-frontend.code-workspace (hunk 2 above)
Comment (apodiptych) [reply wobbly-frontend.code-workspace]: If it's an empty config file that doesn't add any functionality, I'd ignore it -- some workspace settings already live in `.vscode/settings.json`.

File: docker-compose-production.yml (hunk 3 above)
Comment (apodiptych) [reply docker-compose-production.yml]: Apps built with `create-react-native-app` (like this one) use Expo which makes it easy to make native builds for mobile: https://docs.expo.io/versions/v30.0.0/distribution/index.html

For the web, I think we basically just want to statically serve what's in the `public` directory after building with whatever web server we prefer :) That being said, for now, I think we can safely-ish go on with `yarn web` :)

Comment (nam20485) [reply docker-compose-production.yml]: Sounds good. This does yarn web, which will work for now, and we can modify it to suit whatever we want to serve the app with later once we decide.

File: wobbly-frontend.code-workspace (hunk 2 above)
Comment (nam20485) [reply wobbly-frontend.code-workspace]: No it does add functionality, it allows users to launch the VS Code workspace so the source code in the repo is immediately available to edit and develop once pulled. I think this is a significant benefit so that new developers have a powerful and easy to use feature-rich IDE immediately available to develop with, and do not have to set up their own development environment after pulling the source code. Since you have already checked in the .vscode settings, I don't see why you wouldn't want to also check in the workspace file.

**Comment by nam20485** [comment]
Are there any other questions preventing merging this in? It doesn't change anything in your existing process, but does allow me to continue development on the Travis CI pipeline and dockerization. I am in heavy development of those features right now, which is why you see so many commits in the PR. It would be easier to make progress on this if it wasn't stuck in a PR and was in the development branch.

**Comment by KieranPringle** [comment]
@apodiptych my personal preference is to lean on docker's sensible defaults where possible and name files according to convention, but this is consistent with what is in wobbly-backend and I'd say that consistency is more important than convention. 

Maybe we can add some documentation on the steps required to spin up the service in a container? That would help lower the barrier to entry.

File: wobbly-frontend.code-workspace (hunk 2 above)
Comment (apodiptych) [reply wobbly-frontend.code-workspace]: Well, we're assuming they have VS code anyway -- this just adds a shortcut they can click to open the repo (equivalent to running `code .` in the terminal). But this isn't a hill I'm willing to die on ;)

**Comment by marcaaron** [comment]
Should be good to go on this one. But the branch needs to be updated as well? I wonder whether this a necessary constraint since I'm assuming branches will be deleted after they are merged?
